File: 1_Algorithmic_Toolbox/week1_programming_challenges/2_maximum_pairwise_product/max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cassert>

//int MaxPairwiseProduct(const std::vector<int>& numbers) {
//    int max_product = 0;
//    int n = numbers.size();
//
//    for (int first = 0; first < n; ++first) {
//        for (int second = first + 1; second < n; ++second) {
//            max_product = std::max(max_product,
//                numbers[first] * numbers[second]);
//        }
//    }
//
//    return max_product;
//}

int64_t MaxPairwiseProduct(const std::vector<int>& numbers) {
	int max_product = 0;
	int n = numbers.size();
	assert(numbers.size() >= 2);
	int firstMax = std::max(numbers[0], numbers[1]);
	int secondMax = std::min(numbers[0], numbers[1]);
	for (int i = 2;i < numbers.size();++i) {
		secondMax = std::max(secondMax, std::min(firstMax, numbers[i]));
		firstMax = std::max(firstMax, numbers[i]);
	}

	return ((int64_t)firstMax) * ((int64_t)secondMax);
}


int main() {
	int n;
	std::cin >> n;
	std::vector<int> numbers(n);
	for (int i = 0; i < n; ++i) {
		std::cin >> numbers[i];
	}

	std::cout << MaxPairwiseProduct(numbers) << "\n";
	return 0;
}

File: 1_Algorithmic_Toolbox/week1_programming_challenges/2_maximum_pairwise_product/max_pairwise_product.py
def max_pairwise_product(numbers):
    n = len(numbers)
    max_product = 0
    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product,
                numbers[first] * numbers[second])

    return max_product

# ReImplementing max_pairwise_product() using a faster algorithm
def max_pairwise_product_fast(numbers):
    """
    Calculates the maximum pairwise product of a list of numbers.

    Parameters:
    numbers (list): A list of numbers.

    Returns:
    int: The maximum pairwise product of the numbers.

    Examples:
    >>> max_pairwise_product_fast([1, 2, 3])
    6
    >>> max_pairwise_product_fast([7, 5, 14, 2, 8, 8, 10, 1, 2, 3])
    140
    """
    n = len(numbers)
    max_product = 0
    max1 = max(numbers)
    numbers.remove(max1)
    max2 = max(numbers)
    max_product = max1 * max2
    return max_product


if __name__ == '__main__':
    input_n = int(input())
    input_numbers = [int(x) for x in input().split()]
    print(max_pairwise_product(input_numbers))
